tls/s2n_tls13_handshake.c

```diff
@@ -162,91 +162,177 @@ int s2n_tls13_compute_shared_secret(struct s2n_connection *conn, struct s2n_blob
-                server_hs_secret.data, server_hs_secret.size));
+    /* get tls13 key context */
+    s2n_tls13_connection_keys(secrets, conn);
+    bool is_sending_secret = (mode == conn->mode);
+
+    /* produce handshake secret */
+    s2n_stack_blob(hs_secret, secrets.size, S2N_TLS13_SECRET_MAX_LEN);
+
+    uint8_t *finished_data, *implicit_iv_data;
+    struct s2n_session_key *session_key;
+    s2n_secret_type_t secret_type;
+    if (mode == S2N_CLIENT) {
+        finished_data = conn->handshake.client_finished;
+        implicit_iv_data = conn->secure.client_implicit_iv;
+        session_key = &conn->secure.client_key;
+        secret_type = S2N_CLIENT_HANDSHAKE_TRAFFIC_SECRET;
+        conn->client = &conn->secure;
+    } else {
+        finished_data = conn->handshake.server_finished;
+        implicit_iv_data = conn->secure.server_implicit_iv;
+        session_key = &conn->secure.server_key;
+        secret_type = S2N_SERVER_HANDSHAKE_TRAFFIC_SECRET;
+        conn->server = &conn->secure;
     }
 
-    s2n_result_ignore(s2n_key_log_tls13_secret(conn, &client_hs_secret, S2N_CLIENT_HANDSHAKE_TRAFFIC_SECRET));
-    s2n_result_ignore(s2n_key_log_tls13_secret(conn, &server_hs_secret, S2N_SERVER_HANDSHAKE_TRAFFIC_SECRET));
+    struct s2n_hash_state *hash_state;
+    POSIX_GUARD_PTR(hash_state = &conn->handshake.server_hello_copy);
 
-    /* produce handshake traffic keys and configure record algorithm */
-    s2n_tls13_key_blob(server_hs_key, conn->secure.cipher_suite->record_alg->cipher->key_material_size);
-    struct s2n_blob server_hs_iv = { .data = conn->secure.server_implicit_iv, .size = S2N_TLS13_FIXED_IV_LEN };
-    POSIX_GUARD(s2n_tls13_derive_traffic_keys(&secrets, &server_hs_secret, &server_hs_key, &server_hs_iv));
+    POSIX_GUARD(s2n_tls13_derive_handshake_traffic_secret(&secrets, hash_state, &hs_secret, mode));
 
-    s2n_tls13_key_blob(client_hs_key, conn->secure.cipher_suite->record_alg->cipher->key_material_size);
-    struct s2n_blob client_hs_iv = { .data = conn->secure.client_implicit_iv, .size = S2N_TLS13_FIXED_IV_LEN };
-    POSIX_GUARD(s2n_tls13_derive_traffic_keys(&secrets, &client_hs_secret, &client_hs_key, &client_hs_iv));
+    /* trigger secret callbacks */
+    if (conn->secret_cb && conn->config->quic_enabled) {
+        POSIX_GUARD(conn->secret_cb(conn->secret_cb_context, conn, secret_type, hs_secret.data, hs_secret.size));
+    }
+    s2n_result_ignore(s2n_key_log_tls13_secret(conn, &hs_secret, secret_type));
 
-    POSIX_GUARD(conn->secure.cipher_suite->record_alg->cipher->init(&conn->secure.server_key));
-    POSIX_GUARD(conn->secure.cipher_suite->record_alg->cipher->init(&conn->secure.client_key));
+    /* produce handshake traffic keys and configure record algorithm */
+    s2n_tls13_key_blob(hs_key, conn->secure.cipher_suite->record_alg->cipher->key_material_size);
+    struct s2n_blob hs_iv = { .data = implicit_iv_data, .size = S2N_TLS13_FIXED_IV_LEN };
```

Comment:
Is it `S2N_TLS_MAX_IV_LEN == S2N_TLS13_FIXED_IV_LEN` true? Why not the following?
```c
struct s2n_blob hs_iv = {0};
POSIX_GUARD(s2n_blob_init(&hs_iv, implicit_iv_data, S2N_TLS13_FIXED_IV_LEN));
```